### PR TITLE
A15n/lazy render

### PR DESCRIPTION
**Pull request by a15n**
### Overview
This introduces an opt-in for lazy rendering of tooltips and popovers. When `enableLazyRendering=true` a tooltip will only be rendered when the user interacts with the `$target` element *or* when `isShown=true`. This allows tooltip rendering to be delayed until when/if a user needs the tooltip, which results in significant load time improvements for pages with many tooltips and popovers. (see https://github.com/sir-dunxalot/ember-tooltips/issues/100)

### Notable additions
* [lazy-render-wrapper](https://github.com/sir-dunxalot/ember-tooltips/pull/104/files#diff-8096b1ffac251e79ea18b7393d9f241fR1) is a tagless component which will handles rendering and passes through `PASSABLE_OPTIONS`
* [tether-tooltip-and-popover](https://github.com/sir-dunxalot/ember-tooltips/pull/104/files#diff-aab8720cdae20e1f53c656a7aa537184R30) receives the `PASSABLE_OPTIONS` through on `didReceiveAttrs`.
* added new tests
* updated the dummy app and readme

### Inheritance Structure
`lazy-render-wrapper` --extends to--> `tooltip-on-element` --which renders--> `tether-tooltip-on-element`

Popovers follow the same inheritance structure. The `tether-` components lazily render after the user interacts.

### Release Schedule
I recommend we release this as version 2.4.0. This PR keeps `enableLazyRendering=false`. In 3.0.0 we should switch that to `true`. In the meantime people can experiment with lazy rendering on a per-component basis or with [component extension](https://github.com/sir-dunxalot/ember-tooltips#setting-defaults).

### Demo
Note that on page load there is only 1 tooltip/popover. As the user interacts the tooltips are rendered. By the end there are 9 tooltips/popovers in the page.

lazy tooltips

### CODE REVIEW

File: addon/components/tooltip-on-element.js [hunk 1]
```diff
@@ -1,76 +1,54 @@
...
-
-        /* If the event is click, we don't want the
-        click to also trigger focusin */
-
-        if (event !== 'click') {
-          $target.focusin(() => {
-            this.show();
-          });
-        }
+		if (this.get('hasUserInteracted')) {
```
Comment (a15n) [inline addon/components/tooltip-on-element.js]: TODO change this to enableLazyRendering

File: addon/components/lazy-popover-on-element.js [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+export default TooltipAndPopoverComponent.extend({
+
+  /* Options */
+  hideDelay: 250,
+
+  /* Properties */
+  classNames: ['ember-popover'],
+  _isMouseInside: false,
+  didRender() {
```
Comment (a15n) [inline addon/components/lazy-popover-on-element.js]: this is the only change in this file. The rest is a `git mv`.

File: addon/components/lazy-tooltip-on-element.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
+import Ember from 'ember';
```
Comment (a15n) [inline addon/components/lazy-tooltip-on-element.js]: result of a `git mv`

File: tests/integration/components/popover/hover-test.js [hunk 4]
```diff
@@ -33,15 +33,15 @@ test('Popover: hover target, hover elsewhere', function(assert) {
...
-  }, 300);
+  }, 200);
 
   assert.expect(8);
 
 });
 
 test('Popover: hover target, hover popover (too slow)', function(assert) {
 
-  this.render(hbs`{{popover-on-element event="hover" hideDelay="250"}}`);
```
Comment (a15n) [inline tests/integration/components/popover/hover-test.js]: 250 is default

File: addon/components/lazy-popover-on-element.js [hunk 5]
```diff
@@ -0,0 +1,129 @@
+import Ember from 'ember';
```
Comment (a15n) [inline addon/components/lazy-popover-on-element.js]: no changes to this file except the addition of the `didRender`. Appears changed due to `git mv`.

File: addon/components/lazy-tooltip-on-element.js (hunk 3 above)
Comment (a15n) [inline addon/components/lazy-tooltip-on-element.js]: no changes to this file. Appears changed due to `git mv`.

File: tests/integration/components/popover/hover-test.js [hunk 6]
```diff
@@ -33,15 +33,15 @@ test('Popover: hover target, hover elsewhere', function(assert) {
...
+  this.render(hbs`{{popover-on-element event="hover"}}`);
```
Comment (a15n) [inline tests/integration/components/popover/hover-test.js]: 250 is default

**Comment by a15n** [comment]
@sir-dunxalot please review

@MiguelMadero, @SirZach, @lvl99 FYI

**Comment by SirZach** [comment]
That's a lot of code I'm glad I didn't have to write :)

File: README.md [hunk 7]
```diff
@@ -31,6 +32,10 @@ Version 1.0.0 removed <a href="http://darsa.in/tooltip/" target="_blank">darsain
 
 You can use and see the pre-1.0 version on [this branch](https://github.com/sir-dunxalot/ember-tooltips/tree/pre-1.0). Alternatively, install `"ember-tooltips": "0.7.0"` in your `package.json`.
 
+## 2.4.0 Release
+
+Version 2.4.0 introduces lazy rendering. Tooltips and popovers generally don't need to be rendered until the user has interacted with the `$target` element. Adding `enableLazyRendering=true` to your component will enable this future. In version 3.0.0 `enableLazyRendering` will default to `true` and you'll be able to opt-out of lazy rendering as necessary.
```
Comment (sir-dunxalot) [inline README.md]: "will enable this future" <-- Should read "will enable this feature"

File: addon/components/lazy-popover-on-element.js (hunk 5 above)
Comment (sir-dunxalot) [inline addon/components/lazy-popover-on-element.js]: Understood. Thanks for the clarification.

File: addon/components/lazy-popover-on-element.js [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+  classNames: ['ember-popover'],
+  _isMouseInside: false,
+
+  didRender() {
+    // the lazy-render popover component instance needs access to the childView
+    // so that it can call the childView's hide action
+    this._super(...arguments);
+
+    const parentView = this.get('parentView');
+    parentView.set('childView', this);
```
Comment (sir-dunxalot) [inline addon/components/lazy-popover-on-element.js]: I assume there's not a more Emberish way to do this? Maybe by passing an action into the component.

Also, is `childView` too generic here? What si the parent view already has a `childeView` property?

File: addon/components/lazy-render-wrapper.js [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+const { computed, get, run, $ } = Ember;
+
+// https://github.com/emberjs/rfcs/issues/168
+export default function getParent(view) {
+  if (get(view, 'tagName') === '') {
+    // Beware: use of private API! :(
+    if (Ember.ViewUtils && Ember.ViewUtils.getViewBounds) {
+      return $(Ember.ViewUtils.getViewBounds(view).parentElement);
+    } else {
+      return $(view._renderNode.contextualElement);
```
Comment (sir-dunxalot) [inline addon/components/lazy-render-wrapper.js]: How does this code work with the changes incoming from Glimmer (Ember 2.9.0)?

Comment (sir-dunxalot) [inline addon/components/lazy-render-wrapper.js]: Actually, it looks like [the mentioned issue](https://github.com/emberjs/rfcs/issues/168) answers that question. Could you also link to https://github.com/emberjs/ember.js/pull/12500 in the code comment here?

File: addon/components/popover-on-element.js [hunk 10]
```diff
@@ -1,122 +1,14 @@
-import Ember from 'ember';
-import TooltipAndPopoverComponent from 'ember-tooltips/components/tooltip-and-popover';
```
Comment (sir-dunxalot) [inline addon/components/popover-on-element.js]: Am I correct in believing a component no longer exists at this path? If developers have used this component for their own JS or styling then does this create a breaking change?

Comment (sir-dunxalot) [inline addon/components/popover-on-element.js]: I see below that this file exists. Maybe you could just explain how the architecture of the components has changed with this PR? For example, what inherits from what, what wraps what, why existing components can be used outside of the lazy rendering wrapper, etc.

File: addon/components/tooltip-and-popover.js [hunk 11]
```diff
@@ -27,6 +27,16 @@ function cleanNumber(stringOrNumber) {
 
 export default EmberTetherComponent.extend({
 
+  passedPropertiesObject: null,
+  setPropertiesWithpassedPropertiesObject: on('didReceiveAttrs', function() {
```
Comment (sir-dunxalot) [inline addon/components/tooltip-and-popover.js]: Minor point but `passed` is not camelCased in this method name.

File: addon/components/lazy-render-wrapper.js [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+
+		if (this.get('_shouldRender')) {
+			// if the tooltip _shouldRender then we don't need
+			// any special $parent event handling
+			return;
+		}
+
+		const $parent = getParent(this);
+
+		INTERACTION_EVENT_TYPES.forEach((eventType) => {
```
Comment (sir-dunxalot) [inline addon/components/lazy-render-wrapper.js]: Can you please explain how this array is used elsewhere and why.

File: addon/templates/components/popover-on-element.hbs [hunk 13]
```diff
@@ -1,3 +1,12 @@
-{{yield (hash
-	hide=(action "hide")
-)}}
+{{#if _shouldRender}}
```
Comment (sir-dunxalot) [inline addon/templates/components/popover-on-element.hbs]: Correct me if I'm wrong:

- `_shouldRender` will be `false` by default if `enableLazyRendering` is `true`
- `_shouldRender` will be `true` by default if `enableLazyRendering` is `false`

File: tests/integration/components/actions-test.js [hunk 14]
```diff
@@ -12,11 +12,12 @@ test('It supports deprecated lifecycle actions', function(assert) {
   // onTooltip____ actions are deprecated in favor of on on_____ actions
   // these actions will be supported until v3.0.0
 
+  // actions can have unique names
```
Comment (sir-dunxalot) [inline tests/integration/components/actions-test.js]: What's the reason for changing the action names here? (just curious)

File: tests/integration/components/actions-test.js [hunk 15]
```diff
@@ -135,24 +144,45 @@ test('It calls lifecycle actions', function(assert) {
...
 
 });
+
+test('It supports lifecycle closure actions with multiple arguments', function(assert) {
+
+  assert.expect(1);
+
+  let onRenderPassword;
+
+  this.on('onRenderFoo', (trickPassword, realPassword) => {
```
Comment (sir-dunxalot) [inline tests/integration/components/actions-test.js]: Could you add a comment about what this test is doing.

File: tests/integration/components/effect-test.js [hunk 16]
```diff
@@ -0,0 +1,20 @@
+import { moduleForComponent, test } from 'ember-qunit';
```
Comment (sir-dunxalot) [inline tests/integration/components/effect-test.js]: Was this test simply missing before?

File: addon/templates/components/popover-on-element.hbs [hunk 17]
```diff
@@ -1,3 +1,12 @@
...
+	{{#lazy-popover-on-element
```
Comment (sir-dunxalot) [inline addon/templates/components/popover-on-element.hbs]: How does this `lazy-popover-on-element` affect the DOM?

File: README.md [hunk 18]
```diff
@@ -359,6 +373,7 @@ import TooltipOnElementComponent from 'ember-tooltips/components/tooltip-on-elem
 export default TooltipOnElementComponent.extend({
   effect: 'fade',
   side: 'bottom',
+  enableLazyRendering: true,
```
Comment (sir-dunxalot) [inline README.md]: Should this change still be made in `your-app/components/tooltip-on-element}}` as the code example shows?

File: addon/components/popover-on-element.js (hunk 10 above)
Comment (sir-dunxalot) [inline addon/components/popover-on-element.js]: I guess what I'm really trying to get to the bottom of is why we need the new `lazy-render-wrapper` component to achieve lazy loading of components.

**Comment by sir-dunxalot** [comment]
1st review done. Overall this looks like a really sweet addition to the library. I'm very excited to get this in production!

File: README.md (hunk 18 above)
Comment (a15n) [inline README.md]: No, I don't want to switch lazy rendering on by default until version 3.0.0.

I'm just showing it here to illustrate how to opt-in to this change for all of their tooltip/popover components.

File: addon/components/popover-on-element.js (hunk 10 above)
Comment (a15n) [inline addon/components/popover-on-element.js]: I will update the description with a clearer example

File: addon/templates/components/popover-on-element.hbs (hunk 13 above)
Comment (a15n) [inline addon/templates/components/popover-on-element.hbs]: Correct. But let me clarify the first bullet point...

- `_shouldRender` will be `false` by default if `enableLazyRendering` is `true`
  * `_shouldRender` will be `true` by default if `enableLazyRendering=true` **and** `isShown=true`

File: addon/templates/components/popover-on-element.hbs (hunk 17 above)
Comment (a15n) [inline addon/templates/components/popover-on-element.hbs]: The `lazy-popover-on-element` is the Tether component. When this is rendered it will insert the Tether component, which will also attach the tether behavior to the target.

I'm unhappy with the name `lazy-popover-on-element`. I think I'm going to go with `popover-tether-component` and `tooltip-tether-component` to indicate the roles of the various components

File: tests/integration/components/actions-test.js (hunk 14 above)
Comment (a15n) [inline tests/integration/components/actions-test.js]: I added unique action names because when I initially made my changes compliant with this test I wrote a solution dependent on the action names being `onShow`, `onHide`, etc.

File: tests/integration/components/effect-test.js (hunk 16 above)
Comment (a15n) [inline tests/integration/components/effect-test.js]: correct

File: addon/components/lazy-render-wrapper.js (hunk 9 above)
Comment (a15n) [inline addon/components/lazy-render-wrapper.js]: done

File: addon/components/lazy-render-wrapper.js (hunk 12 above)
Comment (a15n) [inline addon/components/lazy-render-wrapper.js]: done

File: tests/integration/components/actions-test.js (hunk 15 above)
Comment (a15n) [inline tests/integration/components/actions-test.js]: done

**Comment by sir-dunxalot** [comment]
@a15n This gets a +1 from me. Are you ok to make releases now? I'm happy to do it but if you want we just need to make sure the following happens:

- [ ] Bump version in `package.json` and commit
- [ ] Create Github tag
- [ ] Push tag to `master`
- [ ] Release on NPM

**Comment by a15n** [comment]
@sir-dunxalot I am ready to release as 2.4.0 right now. I'll update the TODOs you listed as they're completed

**Comment by MiguelMadero** [comment]
You could use ember release to do all of those at once.

(I didnt review the full changes, but Im a little bit worried about the
breaking changes).

**Comment by sir-dunxalot** [comment]
Thanks @MiguelMadero. I use `release-it` locally but that is an interesting option.

I haven't noticed any breaking changes since `2.4.0` - have you seen anything in particular or are just weary because of the number of lines changed?

**Comment by a15n** [comment]
Here is one breaking change, and the PR to address it https://github.com/sir-dunxalot/ember-tooltips/pull/110

File: addon/components/tether-tooltip-and-popover.js [hunk 19]
```diff
@@ -27,6 +27,16 @@ function cleanNumber(stringOrNumber) {
 
 export default EmberTetherComponent.extend({
 
+  passedPropertiesObject: null,
+  setPropertiesWithPassedPropertiesObject: on('didReceiveAttrs', function() {
+    this._super(...arguments);
+
+    let passedPropertiesObject = this.get('passedPropertiesObject');
+    if (passedPropertiesObject) {
+      this.setProperties(passedPropertiesObject);
```
Comment (MiguelMadero) [inline addon/components/tether-tooltip-and-popover.js]: I'm not sure if this is what we want. You're just setting the properties, but not creating a binding. 

let's say we have prop1 as a passedProperty. Then if you change prop1 on the context, it will flow to the target, but it won't flow from the target to the `tether-tooltip-and-popover` since didReceiveAttrs won't be called again, so the actual component will be left with the old value. 

Can you test it? This may not be an issue, but if I'm right, it will be a bad way to break... 

In case it's an issue, it may be as simple as setting adding the same props with a `Binding` suffix or explicitly create bindings between wrapper and wrappee for each property on passedPropertiesObject. 

Also, I'm wondering if it's better to generalize this use case and apply it here, it seems like the "transparent wrapper" or "transparent proxy" has value on its own, but there are also a lot of edge cases that might be easier to test in isolation. I started doing something like this for https://github.com/zenefits/ember-component-apply and something similar and for https://github.com/toranb/ember-cli-hot-loader/pull/25. While both of those approaches work as a "transparent proxy" for components, I don't think they're applicable for ember-tooltips since we rely in one case on transpiling templates on the fly and in the other one to rewrite the AST at buildtime. I only mention those two, because they already cover some of the edge cases for "transparent proxies" and you can look at some of the tests cases for them.

Comment (MiguelMadero) [inline addon/components/tether-tooltip-and-popover.js]: Ok, so this seems like it's not an issue. @andrew-zenefits pointed me to https://github.com/sir-dunxalot/ember-tooltips/pull/104/files#diff-8096b1ffac251e79ea18b7393d9f241fR68 which re-computes the `passedPropertiesObject` when any `PASSABLE_OPTIONS` change.

**Comment by dodeja** [comment]
@a15n you might be already aware of this issue but I want to re-surface it. https://github.com/sir-dunxalot/ember-tooltips/issues/107#issuecomment-261391893

Since IE10 is still a widely used browser I think it would be good to support proper rendering. I will test the issue in other IE browsers and report back.

**Comment by sir-dunxalot** [comment]
Ok, to confirm: #107 is a breaking change and should be released as a `2.x.x` patch.

**Comment by dodeja** [comment]
Visibility: hidden did resolve my issue. Only tested IE 10 though.
